tool/planet/cgroup.go

```diff
@@ -0,0 +1,315 @@
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+/*
+Cgroup Configuration
+--------------------
+Planet uses a customized cgroup structure, that's designed to prevent CPU resource starvation of critical services.
+
+Notes:
+- The cgroup customization is within planet cgroup namespace only
+- Systems with less than 5 cores, will not reserve resources in kubernetes
+  - Relative prioritization will still be applied
+- User tasks will be capped at a maximum CPU usage
+  - 500 millicores on systems with less than 5 cores
+  - 10% of system resources (0.6/6, 1/10, 2/20, 4/40 cores etc) on 6 cores or more
+  - User tasks run with high scheduling priority
+	- THe idea is, an administrator should always be able to troubleshoot a system
+	- However, because CPU usage is capped at 10%, an administrator shouldn't interfere with system services
+- Planet services and user tasks take scheduling priority over kubernetes pods
+  - System and User tasks always have priority over pods
+  - kubernetes remains responsible for setting pod cgroup settings, and relative priority between pods
+*/
+package main
+
+import (
+	"bytes"
+	"fmt"
+	"io/ioutil"
+	"path"
+	"path/filepath"
+	"runtime"
+	"strings"
+	"text/template"
+
+	systemdDbus "github.com/coreos/go-systemd/dbus"
+	"github.com/davecgh/go-spew/spew"
+	"github.com/godbus/dbus"
+
+	"github.com/containerd/cgroups"
+	"github.com/gravitational/planet/lib/constants"
+	"github.com/gravitational/planet/lib/utils"
+	"github.com/gravitational/trace"
+	specs "github.com/opencontainers/runtime-spec/specs-go"
+	"github.com/sirupsen/logrus"
+	"gopkg.in/yaml.v2"
+)
+
+type CgroupConfig struct {
+	// Enabled indicates whether the planet container should apply the cgroup configuration.
+	Enabled bool
+	// Auto indicates whether planet is allowed to regenerate the Cgroup configuration.
+	// If a new version of planet embeds new rules, this allows planet to update the configuration file. If the
+	// config is externally managed or updated by a user, set this to false so planet doesn't overwrite the settings.
+	Auto bool
+
+	// KubeReservedCPU is the amount of CPU to reserve within kubernetes for kubelet + docker (in millicores)
+	KubeReservedCPU int
+	// KubeSystemCPU is the amount of CPU to reserve within kubernetes for system services (in millicores)
+	KubeSystemCPU int
+
+	// Cgroups is a list of cgroup configurations to apply within the planet container
+	Cgroups []CgroupEntry
+}
+
+type CgroupEntry struct {
+	specs.LinuxResources
+
+	// Path is the cgroup hierarchy path this setting applies to
+	Path []string `json:"path"`
```

Comment:
Why does this need a json tag and all other fields don't? Also it would be helpful to document why this is a string slice.